File: test/native/android_emulator_classifier_test.cpp
#include "android_emulator_classifier.h"

#include <iostream>
#include <string>

namespace {

using flutter_defender::platform_internal::AndroidBuildIdentity;
using flutter_defender::platform_internal::IsAndroidEmulator;

AndroidBuildIdentity PhysicalIdentity() {
  return AndroidBuildIdentity{
      "google/husky/husky:15/AP4A.250205.002/1234567:user/release-keys",
      "Pixel 8 Pro",
      "Google",
      "google",
      "husky",
      "husky",
      "husky",
  };
}

bool Expect(const std::string& name, bool actual, bool expected) {
  if (actual == expected) {
    return true;
  }
  std::cerr << name << " failed: expected " << expected << ", got " << actual
            << '\n';
  return false;
}

}  // namespace

int main() {
  bool passed = true;

  AndroidBuildIdentity memu = PhysicalIdentity();
  memu.model = "MEmu";
  passed &= Expect("MEmu identity", IsAndroidEmulator(memu), true);

  AndroidBuildIdentity microvirt = PhysicalIdentity();
  microvirt.manufacturer = "Microvirt";
  passed &= Expect("Microvirt identity", IsAndroidEmulator(microvirt), true);

  AndroidBuildIdentity virtualbox = PhysicalIdentity();
  virtualbox.hardware = "VBOX86";
  passed &= Expect("MEmu VirtualBox hardware", IsAndroidEmulator(virtualbox), true);

  AndroidBuildIdentity avd = PhysicalIdentity();
  avd.fingerprint = "generic/sdk_gphone64_arm64/emu64a:15/test-keys";
  avd.hardware = "ranchu";
  passed &= Expect("Android Virtual Device", IsAndroidEmulator(avd), true);

  passed &= Expect(
      "physical production identity", IsAndroidEmulator(PhysicalIdentity()), false);

  return passed ? 0 : 1;
}

File: android/src/test/kotlin/aleem/flutter/defender/AndroidEmulatorClassifierTest.kt
package aleem.flutter.defender

import kotlin.test.Test
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class AndroidEmulatorClassifierTest {
    @Test
    fun `detects memu model identity`() {
        assertTrue(
            AndroidEmulatorClassifier.isEmulator(
                physicalIdentity.copy(model = "MEmu")
            )
        )
    }

    @Test
    fun `detects microvirt manufacturer identity`() {
        assertTrue(
            AndroidEmulatorClassifier.isEmulator(
                physicalIdentity.copy(manufacturer = "Microvirt")
            )
        )
    }

    @Test
    fun `detects memu virtualbox hardware case insensitively`() {
        assertTrue(
            AndroidEmulatorClassifier.isEmulator(
                physicalIdentity.copy(hardware = "VBOX86")
            )
        )
    }

    @Test
    fun `preserves standard android emulator detection`() {
        assertTrue(
            AndroidEmulatorClassifier.isEmulator(
                physicalIdentity.copy(
                    fingerprint = "generic/sdk_gphone64_arm64/emu64a:15/test-keys",
                    hardware = "ranchu"
                )
            )
        )
    }

    @Test
    fun `does not classify a physical production identity as emulator`() {
        assertFalse(AndroidEmulatorClassifier.isEmulator(physicalIdentity))
    }

    private companion object {
        val physicalIdentity = AndroidBuildIdentity(
            fingerprint = "google/husky/husky:15/AP4A.250205.002/1234567:user/release-keys",
            model = "Pixel 8 Pro",
            manufacturer = "Google",
            brand = "google",
            device = "husky",
            product = "husky",
            hardware = "husky"
        )
    }
}
